Refactor Parser

wrzian opened the pull request:
<details><summary>Alright people, let's do this one last time</summary>
<p>

![Alright people, let's do this one last time (Spiderverse reference)](https://github.com/user-attachments/assets/d4988134-4c54-4e5c-bf0f-b897787ded2c)

</p>
</details>

(This PR is  a cleanup and reorganization of the changes from  #4364)

## Parser Refactor

This is a refactor of Typst's internal parser structure. The grammar of the language itself remains the same, but I've added a new testing structure, greatly expanded the comments and explanations, and I've improved both the ergonomics of the code and the consistency of how we parse trivia tokens (comments and whitespace).

I've numbered the commits and organized them to separate individual changes. Additionally, many of the changes in individual commits are optional or can be locally modified without affecting other things, so the numbers should be helpful for referencing a given change for discussion. (shoutout to `jj` for making this possible and actually ergonomic to manage as opposed to plain git)

None of these changes majorly affects parser performance, but they should serve as a foundation for reasoning about improvements in the future, especially since the `nodes` vector is edited in far fewer places.

### Testing

The first two commits modify the main test runner to add the `--parser-compare` flag for testing the parser.

The first commit just adds the flag, but the second is a major boon, it refactors the existing code so that we can configure the features to avoid recompiling the evaluation pipeline at all if we're only testing the parser. The method is a bit of a hack, but I couldn't come up with a better way (that avoids duplicating most of the code) despite a lot of trying.

The method in (2) is that in `cargo.toml` we declare the main `typst-*` crates optional, but still include them in the "default" feature. Only `typst-syntax` is non-optional and included when compiling in any mode. Thus normal usage is unaffected: the "default" feature is always present. But we can give cargo the `--no-default-features` flag to compile only the `typst-syntax` tests. Then in `test.rs` we add the `#[cfg(feature="default")]` attribute to avoid including parts that would depend on the non-syntax libraries. I tried to keep the annotations as minimal and flexible as possible.

On my machine this takes the test runner compile time from ~20 seconds to ~2 seconds.

The workflow looks like:
```sh
# on commit 2
cargo test --workspace --test tests --no-default-features -- --parser-compare
mkdir _things/main
cp -r /tests/store/syntax _things/main

# on another commit
cargo test --workspace --test tests --no-default-features -- --parser-compare _things/main
```

Since it is a bit of a hack I can understand if you don't want to keep commit (2), but commit (1) should be fine regardless.

### Notable Commits

I won't comment on all of the commits, but here are some things to note:

- **(5-6)** Moved the memoization system to its own area and added some more explicit data structures. This is purely to improve local reasoning about the memoization code (note how (6) is easier to understand since all the changes are local).

- **(7-9)** Returning syntaxnodes from the lexer and parsing raw nodes entirely locally. The main change in (7) seems to be just barely slower, but enables (8) and (9) which make the code so much nicer.

- **(12)** This is the main change to the parser's internal functions and management of trivia. Gives a tiny speedup to undo (7)'s slowdown.

- **(15)** is another major change and unifies markup/code/math mode parsing so all treat trivia the same, where the newline mode is now relevant for each of them, and all newline code is centralized. This also solves the issue with heading spacing from #5138, enabling (17) to finally clean up the stop closures. :grin:

### Syntax tree changes

There are only three actual changes to the output syntax tree, but each is minor and improves consistency.

1. **Commit (14)** In function calls in Math mode, wrapped arguments with trailing trivia no longer wrap that trivia inside the `Math` node, matching how trivia is treated otherwise in math function args. (`maybe_wrap_in_math` change)
    Altered syntax trees: `math-call-2d-repr-structure`, `math-lr-color`, `math-underover-parens`

2. **Commit (15)** Term lists with a newline immediately after the colon now do not include that whitespace in their markup block (matching the current behavior or term lists with no newline immediately after the colon).
    Altered syntax trees: `terms-multiline`

3. **Commit (15)** Trivia nodes in Markup mode now always group together (consistent with Math/Code modes). This mainly manifests in trailing comments in lists or headings now "sticking" to the newline spaces following them. 
    Altered syntax trees: `issue-4632-sth-followed-by-comment`, `heading-trailing-whitespace`

Here is a representation of these syntax tree changes as a diff:
```diff
 1.
-Args[ .., Semicolon(";"), Math[ Text("a"), Space(" ") ], RightParen(")") ]
+Args[ .., Semicolon(";"), Math[ Text("a") ], Space(" "), RightParen(")") ]

 2.
 TermItem[.., Colon(":"), Space("    "), Markup[ Text("item") ] ]
-TermItem[.., Colon(":"), Markup[ Space("\n  "), Text("item") ] ]
+TermItem[.., Colon(":"), Space("\n  "), Markup[ Text("item") ] ]

 3.
- Markup[ .., Heading[ HeadingMarker("="), Markup[BlockComment("/**/") ] ], Space("\n"), .. ]
+ Markup[ .., Heading[ HeadingMarker("="), Markup[] ], BlockComment("/**/"), Space("\n"), .. ]
```

laurmaedje reviewed (commented):
This is really great work! 🎉

It was easy to understand the changes and they make the parser quite a bit easier to follow all around. 

I left comments on the individual commits (unfortunately, GitHub doesn't allow suggestion comments if you're reviewing commit by commit, so I made my suggestions "manually"). 

There are two commits, which I'm not entirely sure about:

- **Commit 15:** I feel like this one, in pursuit of consistency, actually increases complexity a bit through new fields, behaviours, and mechanisms:
  - It adds additional state to the `Token` struct
  - The previously so clean and simple `Parser::lex` function gets quite a bit more complex and the purpose of `eat_preserving_trivia` is also a bit hard to understand
  - The greedy mechanism adds conceptual overhead
  - _Note:_ The `flush_trivia` function is nice and should be kept either way.

  Maybe I just need to understand the idea behind these changes better, but from the get go, I'm not sure it improves things.

- **Commit 18:** This commit feels a bit like abstraction just for the sake of it. It makes the code longer and, in my opinion, not really simpler to grasp. One thing it has going for it, is that it ensures consistency between the functions, but by virtue of `MarkupCommon`'s various fields also only to an extent.

Regarding the **testing stuff:** (A) I can live with the slightly hacky feature flag on `typst-test` if it improves the dev experience when working on the parser. (B) The integration of the parsing stuff into the test runner could be implemented a bit more cleanly in my opinion. As it's not really the focus of this PR, I could imagine tidying that up a bit in a followup PR. Or we could extract it into a preceding PR and I would just iterate a bit on that (refactoring it a little myself might be more productive than me nitpicking around.). I left a few comments on commits 1 & 2 already, but if we defer to a followup PR or split out a separate PR, you could just disregard those and mark them as resolved.

After three paragraphs of stating the few things I _dislike_ here, let me reiterate that I'm very happy with almost everything in this PR. It is really of high quality and the effort you put into splitting things into small, understandable chunks really helped review. Thank you!

wrzian commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1165,8 +1175,11 @@ fn parenthesized_or_array_or_dict(p: &mut Parser) -> SyntaxKind {
 /// State for array/dictionary parsing.
 struct GroupState {
     count: usize,
+    /// A small optimization for if this is _just_ parentheses, `()`.
```
:+1:
I think I was confusing this with the if-statement in `expr_with_paren` for parsing assignments. That if-statement checks that an `=` follows and that the kind is not `Parenthesized`. But this causes us to handle some patterns inconsistently, e.g. both `(x) = 1` and `(_) = 2` give an error in 0.12, but we accept `let (x) = 3`. I assume this is to avoid ambiguity with single length arrays, but it's not fully consistent. Should we change this?

laurmaedje replied:
what change do you propose?

wrzian replied:
I would probably want to disallow `let (x) = {...}` for now and give a hint to add a comma. Even if this is inconsistent with arrow functions, it feels like it would just be subtly incorrect for users trying to deconstruct an array or dict.
Should I make a separate issue to discuss this?

laurmaedje replied:
Yes, let's keep this separate.

wrzian commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1542,52 +1597,68 @@ impl<'s> Parser<'s> {
         self.skip();
     }
 
+    /// Eat the current node then return a reference for in-place mutation.
```
Changing it to `and then`.

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1895,6 +1863,90 @@ impl<'s> Parser<'s> {
+
+    /// Try to load a memoized set of nodes, return None if we did or Some (with
+    /// a checkpoint and a key for the memo map) if we didn't.
+    fn load_memo_or_checkpoint(&mut self) -> Option<(MemoKey, Checkpoint<'s>)> {
```
Sounds good

wrzian commented on crates/typst-syntax/src/lexer.rs:
```diff
@@ -561,11 +564,48 @@ impl Lexer<'_> {
+
+    /// Parse a single MathIdent or an entire FieldAccess SyntaxNode.
+    fn math_ident_or_field(&mut self, start: usize) -> (SyntaxKind, Option<SyntaxNode>) {
+        let initial_ident = self.s.from(start);
+        let Some(mut next_ident) = self.maybe_dot_ident() else {
+            return (SyntaxKind::MathIdent, None);
+        };
+        let mut node = SyntaxNode::leaf(SyntaxKind::MathIdent, initial_ident);
+        loop {
+            let field_children = vec![
+                node,
+                SyntaxNode::leaf(SyntaxKind::Dot, '.'),
+                SyntaxNode::leaf(SyntaxKind::Ident, next_ident),
+            ];
+            node = SyntaxNode::inner(SyntaxKind::FieldAccess, field_children);
+            match self.maybe_dot_ident() {
+                None => break,
+                Some(ident) => next_ident = ident,
+            }
+        }
+        (SyntaxKind::FieldAccess, Some(node))
```
That's a clever rewrite with the while. I like it.

However, I would like to keep `maybe_dot_ident` as a separate function. Mostly just to keep the scanner logic separate properly from the node-building logic, and because it feels like a proper function.

wrzian commented on crates/typst-syntax/src/lexer.rs:
```diff
@@ -561,11 +564,48 @@ impl Lexer<'_> {
+    /// If at a dot and a math identifier, eat and return the identifier.
+    fn maybe_dot_ident(&mut self) -> Option<EcoString> {
+        let start = self.s.cursor();
+        if self.s.eat_if('.') && self.s.at(is_math_id_start) {
```
I think I do like `scout`. For now I'm going with:
```rust
    if self.s.scout(1).is_some_and(is_math_id_start) && self.s.eat_if('.') {
        let ident_start = self.s.cursor();
        self.s.eat();
        self.s.eat_while(is_math_id_continue);
        Some(self.s.from(ident_start))
    } else {
        None
    }
```

Would it be worth adding a new scanner function, `eat_if_both(p1: Pattern, p2: Pattern)` that could solve this pattern (and make use of unsafe code)? It feels like a reasonably general functionality and could be useful for parsing numbers with a single dot, etc.

laurmaedje replied:
I think for now, it wouldn't pull its weight. `scout` is used very rarely so far.
The way you did it is kinda extra clever to avoid the '.' twice.

wrzian commented on tests/suite/math/symbols.typ:
```diff
@@ -0,0 +1,29 @@
+// Test math symbol edge cases.
+
+// --- math-symbol-basic ---
+// #let sym = symbol("s", ("basic", "s"))
+// #test($sym.basic$, $#"s"$)
+
+// --- math-symbol-underscore ---
+// #let sym = symbol("s", ("test_underscore", "s"))
+// // Error: 6-10 unknown symbol modifier
+// $sym.test_underscore$
+
+// --- math-symbol-dash ---
+// #let sym = symbol("s", ("test-dash", "s"))
+// // Error: 6-10 unknown symbol modifier
+// $sym.test-dash$
+
+// --- math-symbol-double ---
+// #let sym = symbol("s", ("test.basic", "s"))
+// #test($sym.test.basic$, $#"s"$)
+
+// --- math-symbol-underscore-double ---
+// #let sym = symbol("s", ("one.test_underscore", "s"))
+// // Error: 10-14 unknown symbol modifier
+// $sym.one.test_underscore$
+
+// --- math-symbol-dash-double ---
+// #let sym = symbol("s", ("one.test-dash", "s"))
+// // Error: 10-14 unknown symbol modifier
+// $sym.one.test-dash$
```
Because I forgot they existed. Uncommenting

wrzian commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1512,27 +1508,37 @@ struct Parser<'s> {
-    /// of the vector.
-    newline_modes: Vec<NewlineMode>,
+    /// A mode for how we proceed on seeing a newline in Code mode.
+    nl_mode: AtNewline,
```
I considered using `nl_stop` for a  bit, but came back to `nl_mode` recently. I'm not sure. I like the renaming of `NewlineMode` to `AtNewline`, but don't have a good name for both the variable holding that state and the `with_*` function that enters it. Thoughts?

laurmaedje replied:
I don't have a good idea either. I think we can just leave it as-is. Though I think `AtNewline` works better for `Continue` and `Stop` than for the other three, since they are not verbs.

wrzian replied:
I changed the names to `ContinueContextual`, `RequireColumn`, and `StopParBreak` on the last rebase. Thoughts?

laurmaedje replied:
Sounds fine

wrzian commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1914,6 +1928,8 @@ impl<'s> Parser<'s> {
     }
 }
 
+// Functions for eating expected or unexpected tokens and generating errors if
```
Oh, neat!

wrzian commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -560,27 +565,27 @@ fn math_args(p: &mut Parser) {
+        // any SyntaxKind::Array elements created in math_args.
+        // (And if we don't follow by wrapping in an array, it has no effect.)
+        // The difference in node layout without this would look like:
+        // - Correct: [ .., Space(" "), Array[Math[],], Semicolon(";"), .. ]
```
```diff
+        // Expression: `$ mat( ;) $`
```

wrzian commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -506,7 +508,8 @@ fn math_args(p: &mut Parser) {
             ";" => {
-                maybe_wrap_in_math(p, arg, named);
+                maybe_wrap_in_math(p, arg, count, named);
+                count = 0;
```
While doing so I noticed something: it looks like `count == 0` implies `namable == true`, except in the case of an unexpected token (although I don't really think there are any unexpected tokens in math). Should we increase count if `p.unexpected()`? If so, should we remove `namable` or leave it for readability?

laurmaedje replied:
There are unexpected tokens in Math: `$ */ $`.

But I think if it's unexpected it doesn't matter much how it's wrapped, since that only affects evaluation, which won't trigger if there are syntax errors.

I think `namable` makes things a bit more readable, so let's keep it.

wrzian commented on tests/src/args.rs:
```diff
@@ -55,6 +57,25 @@ pub struct CliArguments {
+    /// tests/store/syntax/<test>.syntax`
+    #[arg(long)]
+    pub parser_compare: Option<Option<PathBuf>>,
+    // ^ I'm not using a subcommand here because then test patterns don't parse
```
It seems that when (naively) adding a subcommand, CLAP doesn't include other optional positional arguments from the top level. I'm not fully sure, but adding `Parser { compare: Option<PathBuf> }` to the enum `CliArguments` and making `test` take an `Option<Option<PathBuf>>`, I get this behavior:

```
# Basic works
testit parser
testit parser _things/main

# Filtering by name _after_ fails:
testit parser _things/main 'parens'
# => error: unexpected argument 'parens' found

# Filtering by name _before_ doesn't use the command:
testit 'parens' parser _things/main 
# => runs default test suite.
```
I assume there's some fix for this, but despite some short attempts at fixes, I didn't get anything better than this.

laurmaedje replied:
It's okay for the time being.

wrzian commented on tests/Cargo.toml:
```diff
@@ -11,14 +11,32 @@ name = "tests"
-typst-pdf = { workspace = true }
-typst-render = { workspace = true }
-typst-svg = { workspace = true }
+typst-syntax = { workspace = true, optional = false }
```
It's `optional = false`. I chose to keep it to be explicit, but it doesn't really matter. Do you think we should just remove the `optional` on it?

laurmaedje replied:
Yeah, that'd be clearer imo.

laurmaedje replied:
Removing the `optional = false` was marked as resolved, but it doesn't seem like you pushed it.

wrzian replied:
Done.

wrzian commented on tests/src/tests.rs:
```diff
@@ -142,3 +157,42 @@ fn undangle() {
         }
     }
 }
+
```
I added these for now. Agreed on using a different file, but not sure what to call it yet.

wrzian commented on tests/src/args.rs:
```diff
@@ -55,6 +57,25 @@ pub struct CliArguments {
+    ///
+    /// You can generate a correct reference directory by running on a known
+    /// good commit and copying the generated outputs to a new directory.
+    /// `_things` may be a good location as it is in the top-level gitignore.
```
Lol.

PgBiel replied:
One cool trick is `.git/info/exclude` :wink:

wrzian commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -12,54 +12,46 @@ use crate::{ast, set, LexMode, Lexer, SyntaxError, SyntaxKind, SyntaxNode};
-    mut at_start: bool,
-    mut stop: impl FnMut(&Parser) -> bool,
-) {
+fn markup_exprs(p: &mut Parser, mut at_start: bool, stop_set: SyntaxSet) {
```
It does convert what would be a separate check into the single `SyntaxSet::contains` call. Not sure if LLVM makes this meaningless, but I do like how it makes the logic the same in each of the big while-loops (markup/math/code).

laurmaedje replied:
Well, the `End` could be added to the set here before the loop. That's very cheap since set union is just a bitwise OR.

wrzian replied:
See my note on the recent rebase comment: https://github.com/typst/typst/pull/5310#issuecomment-2450410855

wrzian commented:
Upgrades, people, upgrades!

---
Rebased with changes for each of the resolved items. Some more thoughts and larger changes below:

### Commits (1-2)

Added your suggestions here. I think it would work well if you make a PR on this branch in my repo that we then add here as a final merge commit. I.e. you branch off of `parser_refactor` and make a PR onto `wrzian/typst/parser_refactor`, which I accept as a merge becoming commit 19 here.

### Commit (15)

The complexity with `is_greedy`and `eat_preserving_trivia` was entirely to support handling paragraph breaks. The combination let me avoid adding an extra bool to `Token` for whether a newline also contained a paragraph break. This was just a bit too clever for no good reason. I've changed it to just use a bool for `parbreak` now.

Doing so does affect the sytnax tree of `emph-ends-at-parbreak`, but that test is an error, so it shouldn't affect actual parses (and doesn't affect the error message).

On adding extra state to the `Token`, do you think it would be good to reduce the size of the memo map further by only storing the token's previous end and then always re-lexing the token when we restore? It would be a (small) space/speed tradeoff, but does increase complexity a bit.

### Commit (18)

I do like this change, but I'm ok with leaving it out if you think it's not an improvement.

Also, would you prefer if I moved the `with_nl_mode` calls in `markup_expr` into their respective functions (adding an indent to each of them) as in `terms`?

### Other notable changes

- **(3)** I standardized on referring to the modes by their name capitalized, i.e. "in code mode" is now "in Code" everywhere.
    Also, while fixing the comment on parenthesized, I added a new comment to explain the leading colon case for dictionaries - lmk what you think.

- **(9)** Updated math field parsing, uncommented the test file.

- **(14)** Moved the definition of `flush_trivia` here from (15).

- **(15)** Updated the comments on the parser itself with this commit.

- **(17)** I realized that the change here actually wasn't correct with right brackets at the top level, so I added a test for that and fixed it by adding `RightBracket` to all the syntax sets and removing the `break` that would otherwise happen at the top level.

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1484,32 +1514,90 @@ fn pattern_leaf<'s>(
+/// pushing onto the end of the `nodes` vector until a non-trivia kind is found.
+///
+/// The newline mode is used in Code to determine whether a newline should end
+/// the current expression. If so,the parser temporarily changes the current
```
There's a space missing here.

wrzian replied:
Ah yes, I had noticed that on one commit while trying out something else but forgot to fix it

laurmaedje commented:
> I do like this change, but I'm ok with leaving it out if you think it's not an improvement.

I'd prefer to leave it out then.

> Also, would you prefer if I moved the with_nl_mode calls in markup_expr into their respective functions (adding an indent to each of them) as in terms?

Yeah, I think it'd be better if they are moved into the respective functions.

wrzian commented:
Just removing commit 18.

wrzian commented:
Rebased onto main.

The second force-push fixes clippy in commit (12) by allowing the `match_like_matches_macro` lint, since in this case I think it hurts readability, although I would be ok with removing it.

Changes:
- **Commit (15)**
    - Moved the `with_nl_mode` calls into their respective functions.
    - Updated to use a bool as `had_parbreak`.
    - I also pushed a separate branch with a change that creates a `Newline` struct to hold newline info (`column`, `parbreak`), let me know if I should add this or not, see here: https://github.com/wrzian/typst/commit/394ec516a3770b1a6cf5f96733a0386688967a1e 
- **Commit (17)**
    - I forgot to mention that I added a `debug_assert!()` for the syntax sets containing `End`, I think this at least helps the reader better understand the code.
    - I'll also note that this pattern is effectively the same as the `is_terminator()` calls in the various paren handling functions. `is_terminator` is just `matches!(kind, End | Semicolon | RightBrace | RightParen | RightBracket)`.
        Should we use a const SyntaxSet for these or any of the other SyntaxKind `is_*` calls?

laurmaedje commented:
> I forgot to mention that I added a debug_assert!() for the syntax sets containing End, I think this at least helps the reader better understand the code.

If you prefer that over `let set = set.add(SyntaxKind::End)`, I'm okay with it.

> I'll also note that this pattern is effectively the same as the is_terminator() calls in the various paren handling functions. is_terminator is just matches!(kind, End | Semicolon | RightBrace | RightParen | RightBracket).
Should we use a const SyntaxSet for these or any of the other SyntaxKind is_* calls?

You mean remove the `is_` or use it in the `is`? I'd imagine that Rust is better at optimizing a match than the bit thing. (Since LLVM could, after all, lower the match to a bit check). However, it's probably also negligible performance-wise.

laurmaedje commented on Cargo.lock:
It looks like some unrelated lockfile changes slipped in. These should be reverted.

wrzian replied:
Odd. Not sure what happened there. Fixed

laurmaedje commented:
From my side, this looks ready for merge. :)

wrzian commented:
I do prefer the debug asserts, so we can leave that.

And agreed on the performance difference being negligible, I'm good to just leave the `SyntaxKind::is_*` functions  as-is. I did some scrounging around in compiler explorer with a minimized example (<https://godbolt.org/z/v5Ehsbjo4>), but I don't really know what I'm doing, and would need more time to feel confident saying anything about what the compiler does. (I know plenty about the *internals* of CPUs, but that doesn't transfer :upside_down_face:)

I have two outstanding question:

1. Would it be worth decreasing the size of the memo map by re-lexing the current token on exit or by storing the `SyntaxNode` itself in the memo arena, or am I overthinking this? 

2. What are your thoughts on extracting the `column` and `parbreak` into a separate `Newline` struct in commit 15? I think I like it, but I'm not sure if it's just too much hassle. I have a diff at this commit: https://github.com/wrzian/typst/commit/9f8eebaf443a7cde2fa39bfa5248a67b431910d7

laurmaedje commented:
> Would it be worth decreasing the size of the memo map by re-lexing the current token on exit or by storing the SyntaxNode itself in the memo arena, or am I overthinking this?

I think this is also totally negligible performance-wise.

> What are your thoughts on extracting the column and parbreak into a separate Newline struct in commit 15? I think I like it, but I'm not sure if it's just too much hassle. I have a diff at this commit: https://github.com/wrzian/typst/commit/9f8eebaf443a7cde2fa39bfa5248a67b431910d7

Diff looks nice.

wrzian commented:
Second force push is just this change in `current_column()`:
```diff
-        self.token.newline.map_or(0, |newline| newline.column.unwrap_or(0))
+        self.token.newline.and_then(|newline| newline.column).unwrap_or(0)
```

wrzian commented:
Hmm. I went over the behavior of list indentation with whitespace and trivia again, and commit (15) does change the behavior (all versions of it), but none of the tests expose it. I'm restoring the behavior and adding a new test for it as a commit (18).

You should review that, but otherwise I'm good to merge :+1:
